addons/bus/models/bus.py

```diff
@@ -96,38 +96,46 @@ def _gc_messages(self):
 
     @api.model
     def _sendmany(self, notifications):
-        channels = set()
-        values = []
+        _logger.warning("Deprecated method `_sendmany`, use `_sendone` instead")
         for target, notification_type, message in notifications:
-            channel = channel_with_db(self.env.cr.dbname, target)
-            channels.add(channel)
-            values.append({
-                'channel': json_dump(channel),
-                'message': json_dump({
-                    'type': notification_type,
-                    'payload': message,
-                })
-            })
-        self.sudo().create(values)
-        if channels:
+            self._sendone(target, notification_type, message)
+
+    @api.model
+    def _sendone(self, target, notification_type, message):
+        channels = self.env.cr.postcommit.data.setdefault('bus.bus.channels', set())
+        values = self.env.cr.precommit.data.setdefault('bus.bus.values', [])
+        if not channels:
+            @self.env.cr.precommit.add
+            def create_bus():
+                values = self.env.cr.precommit.data['bus.bus.values']
+                self.sudo().create(values)
             # We have to wait until the notifications are commited in database.
             # When calling `NOTIFY imbus`, notifications will be fetched in the
             # bus table. If the transaction is not commited yet, there will be
             # nothing to fetch, and the websocket will return no notification.
             @self.env.cr.postcommit.add
             def notify():
                 with odoo.sql_db.db_connect('postgres').cursor() as cr:
-                    query = sql.SQL("SELECT {}('imbus', %s)").format(sql.Identifier(ODOO_NOTIFY_FUNCTION))
-                    payloads = get_notify_payloads(list(channels))
+                    payloads = get_notify_payloads(list(self.env.cr.postcommit.data['bus.bus.channels']))
                     if len(payloads) > 1:
                         _logger.info("The imbus notification payload was too large, "
                                      "it's been split into %d payloads.", len(payloads))
                     for payload in payloads:
-                        cr.execute(query, (payload,))
-
-    @api.model
-    def _sendone(self, channel, notification_type, message):
-        self._sendmany([[channel, notification_type, message]])
+                        cr.execute(SQL(
+                            "SELECT %s('imbus', %s)",
+                            SQL.identifier(ODOO_NOTIFY_FUNCTION),
+                            payload,
+                        ))
```

Comment:
becoming more than more indent nightmare in this function

Reply:
Using `SQL` is the preferred way to write raw SQL.
I could inline everything, but I don't think it would be more readable.
It would be even far worse if we used automated formatting tools like black though :wink: